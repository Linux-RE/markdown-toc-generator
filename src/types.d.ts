type TocTitle = {
    level: number;
    title: string;
  };
  
  interface TOCOptions {
    format?: 'markdown' | 'html';
    tocTitle?: TocTitle;
    addBackToTop?: boolean;
    numberingStyle?: 'numeric' | 'roman' | 'alphabetical';
    tocFooter?: string;
    maxDepth?: number;
    includeTitles?: string[];
    excludeTitles?: string[];
    indent?: string;
    anchorPrefix?: string;
  }