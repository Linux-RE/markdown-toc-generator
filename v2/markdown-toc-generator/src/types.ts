export interface Heading {
  text: string;
  level: number;
  anchorId?: string;
}

export interface TocItem extends Heading {
  children: TocItem[];
}
