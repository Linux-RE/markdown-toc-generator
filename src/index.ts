<<<<<<< Updated upstream
export function generateToc(markdownContent: string): string {
    const headerPattern = /^(#{1,6})[ \t]+(.+?)[ \t]*$/;
    const toc: string[] = [];

    // Track numbering by header levels
    const numbering: number[] = [];

    // Add the Table of Contents header
    toc.push('## Table of Contents\n');

    const lines = markdownContent.split('\n');

    for (const line of lines) {
        const match = line.match(headerPattern);
        if (match) {
            const level = match[1].length; // Number of '#' defines level
            const title = match[2].trim();

            // Update numbering array
            numbering[level - 1] = (numbering[level - 1] || 0) + 1;
            // Reset deeper levels
            numbering.fill(0, level);

            // Build section number string (e.g., "2.1.3")
            let sectionNumbers = numbering.slice(0, level).join('.');
            if (level === 1) {
                sectionNumbers += '.';
            }

            // Create slug for the link
            const slug = title.toLowerCase().replace(/[^\w\s-]/g, '').replace(/\s+/g, '-');

            // Add proper indentation
            const indent = '   '.repeat(level - 1);

            // Push formatted TOC line
            toc.push(`${indent}${sectionNumbers} [${title}](#${slug})`);
        }
    }

    return toc.join('\n');
}
=======
// types.ts
type TocTitle = {
    level: number; // h1, h2, etc.
    title: string;
};

interface TOCOptions {
    format?: 'markdown' | 'html';
    tocTitle?: TocTitle;
    addBackToTop?: boolean;
    numberingStyle?: 'numeric' | 'roman' | 'alphabetical';
    tocFooter?: string;
    maxDepth?: number;
    includeTitles?: string[];
    excludeTitles?: string[];
    indent?: string;
    anchorPrefix?: string;
}

// index.ts
export function generateTOC(markdownContent: string, options: TOCOptions = {}): string {
    const {
        format = 'markdown',
        tocTitle,
        addBackToTop = false,
        numberingStyle = 'numeric',
        tocFooter,
        maxDepth = 6,
        includeTitles,
        excludeTitles,
        indent = '  ',
        anchorPrefix = '',
    } = options;

    const lines = markdownContent.split('\n');
    const numbering: number[] = [];
    const tocLines: string[] = [];

    const formatSlug = (title: string) => {
        return anchorPrefix + title
            .toLowerCase()
            .replace(/[^\w\s-]/g, '')
            .trim()
            .replace(/\s+/g, '-')
            .replace(/-+/g, '-');
    };

    const getNumbering = (level: number) => {
        numbering[level - 1] = (numbering[level - 1] || 0) + 1;
        numbering.fill(0, level);
        return numbering.slice(0, level).filter(n => n > 0);
    };

    const formatNumber = (numbers: number[]) => {
        if (numberingStyle === 'roman') {
            const toRoman = (n: number) => {
                const romans: { [key: string]: number } = {
                    M: 1000,
                    CM: 900,
                    D: 500,
                    CD: 400,
                    C: 100,
                    XC: 90,
                    L: 50,
                    XL: 40,
                    X: 10,
                    IX: 9,
                    V: 5,
                    IV: 4,
                    I: 1,
                };
                let result = '';
                for (const key in romans) {
                    const value = romans[key];
                    while (n >= value) {
                        result += key;
                        n -= value;
                    }
                }
                return result.toLowerCase();
            };
            return numbers.map(toRoman).join('.');
        }
        if (numberingStyle === 'alphabetical') {
            const toAlpha = (n: number) => String.fromCharCode(96 + n);
            return numbers.map(toAlpha).join('.');
        }
        // Default numeric
        return numbers.join('.');
    };

    if (format === 'html') {
        tocLines.push('<nav aria-label="Table of Contents">');

        if (tocTitle) {
            const titleTag = `h${tocTitle.level || 2}`;
            tocLines.push(`  <${titleTag}>${tocTitle.title}</${titleTag}>`);
        }

        tocLines.push('  <ol>');
    } else {
        if (tocTitle) {
            const hashes = '#'.repeat(tocTitle.level || 2);
            tocLines.push(`${hashes} ${tocTitle.title}`);
            tocLines.push('');
        }
    }

    const stack: string[] = [];

    lines.forEach(line => {
        const trimmed = line.trimStart();
        if (!trimmed.startsWith('#')) return;

        let level = 0;
        while (trimmed[level] === '#') {
            level++;
        }

        if (level < 1 || level > 6 || level > maxDepth) return;

        const title = trimmed.slice(level).trim();
        if (!title) return;

        if (includeTitles && !includeTitles.includes(title)) return;
        if (excludeTitles && excludeTitles.includes(title)) return;

        const slug = formatSlug(title);
        const numberParts = getNumbering(level);
        const sectionNumber = formatNumber(numberParts);

        if (format === 'html') {
            while (stack.length >= level) {
                tocLines.push(`${'  '.repeat(stack.length + 1)}</ol>`);
                tocLines.push(`${'  '.repeat(stack.length)}</li>`);
                stack.pop();
            }

            tocLines.push(`${'  '.repeat(level)}<li>`);
            tocLines.push(`${'  '.repeat(level + 1)}<a href="#${slug}">${sectionNumber} ${title}</a>`);
            tocLines.push(`${'  '.repeat(level + 1)}<ol>`);
            stack.push('ol');
        } else {
            const line = `${indent.repeat(level - 1)}- [${sectionNumber} ${title}](#${slug})`;
            tocLines.push(line);
        }
    });

    if (format === 'html') {
        while (stack.length) {
            tocLines.push(`${'  '.repeat(stack.length + 1)}</ol>`);
            tocLines.push(`${'  '.repeat(stack.length)}</li>`);
            stack.pop();
        }

        if (addBackToTop) {
            tocLines.push('    <li><a href="#">Back to Top</a></li>');
        }

        tocLines.push('  </ol>');

        if (tocFooter) {
            tocLines.push(`  <p><em>${tocFooter}</em></p>`);
        }

        tocLines.push('</nav>');
    } else {
        if (addBackToTop) {
            tocLines.push('');
            tocLines.push('- [Back to Top](#)');
        }

        if (tocFooter) {
            tocLines.push('');
            tocLines.push(`*${tocFooter}*`);
        }
    }

    return tocLines.join('\n');
}

const markdownContent = `
# Introduction
## Getting Started
### Requirements
## Installation
### Using npm
# Usage
# FAQ
`;

const toc = generateTOC(markdownContent, {
    format: 'markdown',
    tocTitle: { level: 1, title: 'My Custom Table of Contents' },
    addBackToTop: true,
    numberingStyle: 'numeric',
    tocFooter: 'Generated by My Tool',
    anchorPrefix: 'docs/',
});

console.log(toc);
>>>>>>> Stashed changes
