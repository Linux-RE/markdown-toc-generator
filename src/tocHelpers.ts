export const getOrderedListType = (style: string): string => {
    switch (style) {
        case 'roman': return 'I';
        case 'alphabetical': return 'A';
        case 'numeric':
        default: return '1';
    }
};

export const generateSlug = (title: string): string =>
    title.toLowerCase()
        .replace(/[^\w\s-]/g, '')
        .trim()
        .replace(/\s+/g, '-')
        .replace(/-+/g, '-');

export const toRoman = (num: number): string => {
    const romans: [number, string][] = [
        [1000, 'M'], [900, 'CM'], [500, 'D'], [400, 'CD'],
        [100, 'C'], [90, 'XC'], [50, 'L'], [40, 'XL'],
        [10, 'X'], [9, 'IX'], [5, 'V'], [4, 'IV'], [1, 'I']
    ];
    let res = '';
    for (const [val, sym] of romans) {
        while (num >= val) { res += sym; num -= val; }
    }
    return res;
};
