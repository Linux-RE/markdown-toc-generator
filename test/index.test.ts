import { generateToc } from '../src/index';

describe('generateToc', () => {
  it('should generate correct TOC for a basic Markdown input', () => {
    const markdown = `
# Introduction

## Getting Started  
### Requirements  
### Quick Start  

## Installation  
### Using npm  
### Using Docker  

## Usage  
### Basic Example  
### Advanced Configuration  

## API Reference  
### Authentication  
### Endpoints  
### Error Codes  

## Examples  

## Troubleshooting  

## FAQ  

## License
`;

    const expectedToc = `
## Table of Contents

1. [Introduction](#introduction)
2. [Getting Started](#getting-started)
   2.1 [Requirements](#requirements)
   2.2 [Quick Start](#quick-start)
3. [Installation](#installation)
   3.1 [Using npm](#using-npm)
   3.2 [Using Docker](#using-docker)
4. [Usage](#usage)
   4.1 [Basic Example](#basic-example)
   4.2 [Advanced Configuration](#advanced-configuration)
5. [API Reference](#api-reference)
   5.1 [Authentication](#authentication)
   5.2 [Endpoints](#endpoints)
   5.3 [Error Codes](#error-codes)
6. [Examples](#examples)
7. [Troubleshooting](#troubleshooting)
8. [FAQ](#faq)
9. [License](#license)
`.trim();

    const result = generateToc(markdown);
    expect(result).toBe(expectedToc);
  });

  // Other tests can be added here...
});
