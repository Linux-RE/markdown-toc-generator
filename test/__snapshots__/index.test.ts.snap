// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateNestedToc should generate a bulleted Table of Contents 1`] = `
"## Table of Contents

- [Introduction](#introduction)
   - [Getting Started](#getting-started)
      - [Requirements](#requirements)
      - [Quick Start](#quick-start)
   - [Installation](#installation)
      - [Using npm](#using-npm)
      - [Using Docker](#using-docker)
   - [Usage](#usage)
      - [Basic Example](#basic-example)
      - [Advanced Configuration](#advanced-configuration)
   - [API Reference](#api-reference)
      - [Authentication](#authentication)
      - [Endpoints](#endpoints)
      - [Error Codes](#error-codes)
   - [Examples](#examples)
   - [Troubleshooting](#troubleshooting)
   - [FAQ](#faq)
   - [License](#license)"
`;

exports[`generateNestedToc should generate a numbered Table of Contents 1`] = `
"## Table of Contents

1. [Introduction](#introduction)
   1.1. [Getting Started](#getting-started)
      1.1.1. [Requirements](#requirements)
      1.1.2. [Quick Start](#quick-start)
   1.2. [Installation](#installation)
      1.2.1. [Using npm](#using-npm)
      1.2.2. [Using Docker](#using-docker)
   1.3. [Usage](#usage)
      1.3.1. [Basic Example](#basic-example)
      1.3.2. [Advanced Configuration](#advanced-configuration)
   1.4. [API Reference](#api-reference)
      1.4.1. [Authentication](#authentication)
      1.4.2. [Endpoints](#endpoints)
      1.4.3. [Error Codes](#error-codes)
   1.5. [Examples](#examples)
   1.6. [Troubleshooting](#troubleshooting)
   1.7. [FAQ](#faq)
   1.8. [License](#license)"
`;

exports[`generateTOC should generate a bulleted Table of Contents 1`] = `"## Table of Contents"`;

exports[`generateTOC should generate a numbered Table of Contents 1`] = `"## Table of Contents"`;
